File: tests/day10.rs
mod common;

use std::str::FromStr;

fn read() -> Vec<Vec<Delim>> {
    let parsed: Vec<Vec<Delim>> = common::read_test_input("data/day-10/input.txt")
        .iter()
        .map(|line_s: &String| -> Vec<Delim> {
            line_s.split("").flat_map(|c| c.parse().ok()).collect()
        })
        .collect();
    return parsed;
}

enum SyntaxCheckResult {
    Incomplete { score: usize },
    Corrupt { score: usize },
}

impl SyntaxCheckResult {
    fn corrupt_score(&self) -> usize {
        return match *self {
            SyntaxCheckResult::Corrupt { score } => score,
            _ => 0,
        };
    }
    fn complete_score(&self) -> usize {
        return match *self {
            SyntaxCheckResult::Incomplete { score } => score,
            _ => 0,
        };
    }
}

#[derive(Copy, Clone, Debug, PartialEq)]
enum Delim {
    LRound,
    LSquare,
    LCurly,
    LAngle,
    RRound,
    RSquare,
    RCurly,
    RAngle,
}

use Delim::*;
impl FromStr for Delim {
    type Err = ();

    fn from_str(input: &str) -> Result<Delim, Self::Err> {
        match input {
            "(" => Ok(LRound),
            "[" => Ok(LSquare),
            "{" => Ok(LCurly),
            "<" => Ok(LAngle),
            ")" => Ok(RRound),
            "]" => Ok(RSquare),
            "}" => Ok(RCurly),
            ">" => Ok(RAngle),
            _ => Err(()),
        }
    }
}

impl Delim {
    fn is_open(&self) -> bool {
        return *self == LRound || *self == LSquare || *self == LCurly || *self == LAngle;
    }

    fn is_close(&self) -> bool {
        return !self.is_open();
    }

    fn matches(&self, other: &Delim) -> bool {
        return (self.is_close() && other.matches(self))
            || *self == LRound && *other == RRound
            || *self == LSquare && *other == RSquare
            || *self == LCurly && *other == RCurly
            || *self == LAngle && *other == RAngle;
    }

    fn to_closing(&self) -> Option<Delim> {
        return match *self {
            LRound => Some(RRound),
            LSquare => Some(RSquare),
            LCurly => Some(RCurly),
            LAngle => Some(RAngle),
            _ => None,
        };
    }

    fn corrupt_score(&self) -> usize {
        return match self {
            RRound => 3,
            RSquare => 57,
            RCurly => 1197,
            RAngle => 25137,
            _ => 0,
        };
    }

    fn close_score(&self) -> usize {
        return match self {
            RRound => 1,
            RSquare => 2,
            RCurly => 3,
            RAngle => 4,
            _ => 0,
        };
    }
}

fn check_syntax(line: &Vec<Delim>) -> SyntaxCheckResult {
    let mut stack: Vec<Delim> = Vec::new();
    for delim in line {
        if delim.is_open() {
            stack.push(*delim);
        } else {
            if let Some(last) = stack.last() {
                if last.matches(delim) {
                    stack.pop();
                } else {
                    return SyntaxCheckResult::Corrupt {
                        score: delim.corrupt_score(),
                    };
                }
            }
        }
    }
    let to_close: Vec<Delim> = stack.iter().rev().flat_map(|d| d.to_closing()).collect();
    let close_score = to_close.iter().fold(0, |a, v| (5 * a) + v.close_score());
    SyntaxCheckResult::Incomplete { score: close_score }
}

#[test]
fn part1() {
    let input = read();

    assert_eq!(90, input.len(), "expect number of lines");

    let results: Vec<SyntaxCheckResult> = input.iter().map(|line| check_syntax(line)).collect();

    assert_eq!(90, results.len(), "expect number of results");

    let corrupt_score: usize = results
        .iter()
        .map(|r| r.corrupt_score())
        .fold(0, |a, v| a + v);

    assert_eq!(339411, corrupt_score, "expect corrupt score");
}

#[test]
fn part2() {
    let input = read();

    assert_eq!(90, input.len(), "expect number of lines");

    let results: Vec<SyntaxCheckResult> = input.iter().map(|line| check_syntax(line)).collect();

    assert_eq!(90, results.len(), "expect number of results");

    let mut complete_scores: Vec<usize> = results
        .iter()
        .map(|r| r.complete_score())
        .filter(|&s| s > 0)
        .collect();

    complete_scores.sort();

    let complete_score = complete_scores[complete_scores.len() / 2];

    assert_ne!(741714, complete_score, "too low");
    assert_ne!(316965209766, complete_score, "too high");
    assert_eq!(2289754624, complete_score, "expect complete score");
}

File: ts-src/day10.ts
import { map, Observable, reduce } from "rxjs";
import { readTestInputLines } from "./util";

export enum SyntaxCheckResultKind {
    Incomplete,
    Corrupt,
}

export type SyntaxCheckResult = { kind: SyntaxCheckResultKind, score: number };

export enum Delim {
    LRound,
    LSquare,
    LCurly,
    LAngle,
    RRound,
    RSquare,
    RCurly,
    RAngle,
}

const parseDelim = (c: string): Delim[] => {
    switch (c) {
        case '(': return [Delim.LRound];
        case '[': return [Delim.LSquare];
        case '{': return [Delim.LCurly];
        case '<': return [Delim.LAngle];
        case ')': return [Delim.RRound];
        case ']': return [Delim.RSquare];
        case '}': return [Delim.RCurly];
        case '>': return [Delim.RAngle];
        default: return [];
    }
};

export const isOpen = (delim: Delim): boolean => {
    switch (delim) {
        case Delim.LRound:
        case Delim.LSquare:
        case Delim.LCurly:
        case Delim.LAngle: return true;
        default: return false;
    }
}

export const isClose = (delim: Delim): boolean => {
    return !isOpen(delim);
}

export const matches = (left: Delim, right: Delim): boolean => {
    return (isClose(left) && matches(right, left))
        || left == Delim.LRound && right == Delim.RRound
        || left == Delim.LSquare && right == Delim.RSquare
        || left == Delim.LCurly && right == Delim.RCurly
        || left == Delim.LAngle && right == Delim.RAngle;
}

export const toClosing = (delim: Delim): Delim => {
    switch (delim) {
        case Delim.LRound: return Delim.RRound;
        case Delim.LSquare: return Delim.RSquare;
        case Delim.LCurly: return Delim.RCurly;
        case Delim.LAngle: return Delim.RAngle;
        default: return delim;
    }
}

export const corruptScore = (delim: Delim): number => {
    switch (delim) {
        case Delim.RRound: return 3;
        case Delim.RSquare: return 57;
        case Delim.RCurly: return 1197;
        case Delim.RAngle: return 25137;
        default: return 0;
    }
}

export const closeScore = (delim: Delim): number => {
    switch (delim) {
        case Delim.RRound: return 1;
        case Delim.RSquare: return 2;
        case Delim.RCurly: return 3;
        case Delim.RAngle: return 4;
        default: return 0;
    }
}

export const checkSyntax = (line: Delim[]): SyntaxCheckResult => {
    var stack: Delim[] = [];
    for (var i = 0; i < line.length; i++) {
        var delim = line[i];
        if (isOpen(delim)) {
            stack.push(delim);
        } else {
            if (stack.length > 0) {
                if (matches(stack.slice(-1)[0], delim)) {
                    stack.pop();
                } else {
                    return { kind: SyntaxCheckResultKind.Corrupt, score: corruptScore(delim) };
                }
            }
        }
    }
    return {
        score: stack.reverse().map(delim => toClosing(delim)).reduce((acc, value) => (5 * acc) + closeScore(value), 0),
        kind: SyntaxCheckResultKind.Incomplete
    };
}

export const readInput = (): Observable<Delim[]> => readTestInputLines('day-10').pipe(
    map(line => line.split('').flatMap(c => parseDelim(c))),
);
